coderd/pointer/pointer.go

```diff
@@ -0,0 +1,38 @@
+package pointer
+
+import (
+	"context"
+
+	"go.uber.org/atomic"
+)
+
+// New constructs a Handle with an initialized value.
+func New[T any](value T) *Handle[T] {
+	h := &Handle[T]{
+		key: struct{}{},
+		ptr: atomic.Pointer[T]{},
+	}
+	h.Store(value)
+	return h
+}
+
+// Handle loads the stored value into a context, and returns
+// a context with the attached value. It's intention is to
+// hold a single handle for the lifecycle of a request.
+type Handle[T any] struct {
+	key struct{}
+	ptr atomic.Pointer[T]
+}
+
+func (p *Handle[T]) Load(ctx context.Context) (context.Context, T) {
+	value, ok := ctx.Value(&p.key).(T)
+	if !ok {
+		ctx = context.WithValue(ctx, &p.key, *p.ptr.Load())
+		return p.Load(ctx)
+	}
+	return ctx, value
+}
+
+func (p *Handle[T]) Store(t T) {
```

Comment:
Unfortunately, this still leaves the `Store()` functionality accessible to any code that can access the `Handle`, which in the current case is any code with a reference to the AGPL `*API`.

Reply:
I'm not convinced that's a big deal. It's pretty obvious to a caller that they shouldn't override an implementation from my perspective.